File: templates/dart-shelf/lib/helpers/mysql/lib/meta.dart
enum RelationKey { AND, OR }

enum OrderKey {
  ASC,
  DESC,
}

enum Condition {
  eq(value: '='),
  ne(value: '!='),
  gt(value: '>'),
  gte(value: '>='),
  lt(value: '<'),
  lte(value: '<='),
  like(value: 'LIKE'),
  include(value: 'IN'),
  area(value: 'BETWEEN'), // 范围
  not(value: '='); // 'NOT {name} = {value}'

  const Condition({required this.value});
  final String value;
}

class SqlCondition {
  RelationKey relation;
  String field;
  dynamic value;
  Condition condition;
  SqlCondition(this.field, this.value,
      {this.condition = Condition.eq, this.relation = RelationKey.AND});

  String get valueSql {
    String target = '';
    if (condition == Condition.include) {
      if (value is List && value.isNotEmpty) {
        target = MetaUtil.valuesLinkTo((value));
      }
    } else if (condition == Condition.area) {
      if (value is List && value.length > 1) {
        target = '${value[0]} AND ${value[1]}';
      }
    } else {
      target = '${MetaUtil.valueTypeOf(value)}';
    }
    return target;
  }

  String get sql {
    String target = '';
    if (valueSql.isNotEmpty) {
      if (condition == Condition.not) {
        target += 'NOT ';
      }
      target += '$field ${condition.value} $valueSql';
    }
    return target;
  }

  link(String prev) {
    if (prev.isEmpty) {
      return sql;
    }
    if (sql.isEmpty) {
      return prev;
    }
    return '$prev ${relation.name} $sql';
  }
}

class SqlOptions {
  int limit = 0;
  int skip = 0;
  // 排序
  Map<String, OrderKey> order = {};

  get sql {
    String target = '';
    if (order.isNotEmpty) {
      List<String> items = [];
      order.forEach((name, v) {
        items.add('$name ${v.name}');
      });
      target += 'ORDER BY ${items.join(', ')}';
    }
    if (limit > 0) {
      target += '\nLIMIT $limit OFFSET $skip';
    }
    return target;
  }
}

class MetaUtil {
  static valueTypeOf(dynamic v) {
    if (v is String) {
      return "'$v'";
    }
    return v;
  }

  static valuesLinkTo(List<dynamic> values) {
    List<dynamic> items = [];
    for (var v in values) {
      items.add(valueTypeOf(v));
    }
    return '(${items.join(', ')})';
  }

  static conditionToSql(List<SqlCondition> list) {
    String target = '';
    for (SqlCondition ele in list) {
      target = ele.link(target);
    }
    return '$target\n';
  }
}

File: templates/node-koa/src/helpers/mysql/lib/meta.ts
import { MysqlParser } from './utils';

export class Condition {
  relation: RelationKey;
  field: string;
  value: unknown;
  condition: ConditionKey;
  constructor(props: ConditionProps) {
    this.field = props.field;
    this.value = props.value;
    this.condition = props.condition || ConditionKey.eq;
    this.relation = props.relation || RelationKey.and;
  }

  get #valueSql() {
    let target: string = '';
    const { condition, value } = this;
    if (condition == ConditionKey.include) {
      if (Array.isArray(value) && value.length > 0) {
        target = MysqlParser.toValuesOf(value);
      }
    } else if (condition == ConditionKey.area) {
      if (Array.isArray(value) && value.length > 0) {
        target = `${value[0]} AND ${value[1]}`;
      }
    } else {
      target = MysqlParser.valueTypeOf(value);
    }
    return target;
  }

  get result() {
    let target: string = '';
    const { condition, field } = this;
    if (this.#valueSql) {
      if (condition == ConditionKey.not) {
        target += 'NOT ';
      }
      target += `${field} ${condition} ${this.#valueSql}`;
    }
    return target;
  }

  connect(prev: string) {
    const { relation, result } = this;
    if (!prev) {
      return result;
    }
    if (!result) {
      return prev;
    }
    return `${prev} ${relation} ${result}`;
  }
}

export class Options {
  limit: number = 0;
  skip: number = 0;
  // 排序
  sort: Record<string, SortKey> = {};

  get result() {
    let target: string = '';
    const { sort, limit, skip } = this;
    if (Object.keys(sort).length > 0) {
      const items: string[] = [];
      Object.keys(sort).forEach((name, v) => {
        items.push(`${name} ${v}`);
      });
      target += `ORDER BY ${items.join('), ')}`;
    }
    if (limit > 0) {
      target += `\nLIMIT ${limit} OFFSET ${skip}`;
    }
    return target;
  }
}

export enum RelationKey {
  and = 'AND',
  or = 'OR',
}

export enum SortKey {
  asc = 'ASC',
  desc = 'DESC',
}

export enum ConditionKey {
  eq = '=',
  ne = '!=',
  gt = '>',
  gte = '>=',
  lt = '<',
  lte = '<=',
  like = 'LIKE',
  include = 'IN',
  area = 'BETWEEN', // 范围
  not = '=', // 'NOT {name} = {value}'
}

export interface ConditionProps {
  field: string;
  value: unknown;
  relation?: RelationKey;
  condition?: ConditionKey;
}
